modules/stateful/main.tf

```diff
@@ -19,6 +19,13 @@ resource "aws_volume_attachment" "default" {
   device_name = "/dev/xvdh"
   volume_id   = "${element(aws_ebs_volume.default.*.id, count.index)}"
   instance_id = "${element(aws_instance.default.*.id, count.index)}"
+
+  # There's a circular dependency here on destroy with the instance. In order
+  # to dettach, you must first unmount. We assume that the attachment would only
+  # be destroyed when the instance is also being destoryed. In that case, we can
+  # skip destroy of the attachment, and just remove it from state. Then the
+  # instance will be destroyed, removing the attachment.
+  skip_destroy = true
```

Comment:
So we want to keep the volume. Just curious, why it failed to be destroyed when instance is destroyed, is it because the volume in created by other process?

Reply:
I should've linked to some of the issues I found. It's sort of a circular dependency and a [bug in Terraform](https://github.com/terraform-providers/terraform-provider-aws/issues/6024).

The aws_volume_attachment depends on the aws_instance. So when you start a replace or destroy on the aws_instance, the aws_volume_attachment starts the destroy process first. But since the EBS volume is still _mounted_ within the EC2 instance, the aws_volume_attachment times out. Terraform can't actually handle the unmount. I saw a few work arounds that include doing a shutdown on the ec2 instance as part of the aws_volume_attachment destroy, but using skip_destroy seems to be the simplest work around.